src/persist-client/src/internal/machine.rs

```diff
@@ -361,11 +361,36 @@ where
         let mut machine = self;
         spawn(|| "persist::heartbeat_read", async move {
             let sleep_duration = machine.cfg.reader_lease_duration / 2;
+            let mut last_wakeup = Instant::now();
             loop {
                 tokio::time::sleep(sleep_duration).await;
+                let wakeup = Instant::now();
+
+                let elapsed_since_last_wakeup = wakeup - last_wakeup;
+                if elapsed_since_last_wakeup > machine.cfg.reader_lease_duration {
+                    warn!(
+                        "reader ({}) of shard ({}) went {}s between heartbeats",
+                        reader_id,
+                        machine.shard_id(),
+                        elapsed_since_last_wakeup.as_secs_f64()
+                    );
+                }
+
                 let (_seqno, existed, _maintenance) = machine
                     .heartbeat_reader(&reader_id, (machine.cfg.now)())
                     .await;
+
+                let elapsed_since_this_wakeup = Instant::now() - wakeup;
+                if elapsed_since_this_wakeup > Duration::from_secs(60) {
+                    warn!(
+                        "reader ({}) of shard ({}) heartbeat call took {}s",
+                        reader_id,
+                        machine.shard_id(),
+                        elapsed_since_this_wakeup.as_secs_f64(),
+                    );
+                }
+
+                last_wakeup = wakeup;
```

Comment:
it might be cleaner to just make an `Instant` before the `sleep`, and then `tht_instant.elapsed()` after it, and similarly for the length of the `heartbeat_reader` call, instead of reasoning about a variable changing